File: mce/custom/purchase_receipt/purchase_receipt.py

import frappe 
import json


@frappe.whitelist()
def get_fines_names(self): 
    self = frappe._dict(json.loads(self))
    bo_list = list()
    for i in self.get('items'):
        if i.get('purchase_order'): 
            po_doc = frappe.get_doc('Purchase Order' ,i.get('purchase_order') )
            if po_doc.custom_blanket_order: 
                bo_list.append(po_doc.custom_blanket_order)
    bo_list = list(set(bo_list))
    fines_list = list()
    for bo in bo_list:
        bo_doc = frappe.get_doc('Blanket Order' , bo)
        if len(bo_doc.custom_fines) != 0: 
            for f in bo_doc.custom_fines: 
                if f.purchase_fines not in fines_list: 
                    fines_list.append(f.purchase_fines)
    return fines_list

File: mce/custom/purchase_receipt/purchase_receipt.js
frappe.ui.form.on('Purchase Receipt', {
    setup: function(frm) {
        Filter(frm);
    },
    refresh: function(frm) {
        Filter(frm);
    },
    onload: function(frm) {
        Filter(frm);
    },
});

frappe.ui.form.on('Fines Details', {
    form_render: function(frm) {
        Filter(frm);
    },
});

function Filter(frm) {
    frappe.call({
        method: "mce.custom.purchase_receipt.purchase_receipt.get_fines_names",
        args: {
            self: frm.doc
        },
        callback: function(r) {
            if (r.message) {
                const finesList = r.message;
                console.log(finesList);
                frm.fields_dict["custom_fines"].grid.get_field("purchase_fines").get_query = function(doc) {
                    return {
                        filters: {
                            "fines_name": ["in", finesList]
                        }
                    }
                }
            }
            
        }
    });
}